src/circuit_proof/mod.rs

```diff
@@ -0,0 +1,781 @@
+
+        // Get IPP variables
+        let (x_sq, x_inv_sq, s) = self.ipp_proof.verification_scalars(transcript);
+        let s_inv = s.iter().rev().take(circuit.n);
+        let a = self.ipp_proof.a;
+        let b = self.ipp_proof.b;
+
+        // define parameters for P check
+        let g = s.iter().take(circuit.n).map(|s_i| -a * s_i);
+        let h = s_inv
+            .zip(util::exp_iter(y.invert()))
+            .map(|(s_i_inv, exp_y_inv)| -exp_y_inv * b * s_i_inv - Scalar::one());
+
+        // define parameters for t check
+        let delta = inner_product(&W_R_flatten_yinv, &W_L_flatten);
+        let powers_of_z: Vec<Scalar> = util::exp_iter(z).take(circuit.q).collect();
+        let z_c = z * inner_product(&powers_of_z, &circuit.c);
+        let W_V_flatten: Vec<Scalar> = matrix_flatten(circuit.W_V, z, circuit.m)?;
+        let V_multiplier = W_V_flatten.iter().map(|W_V_i| r * x * x * W_V_i);
+
+        let mega_check = RistrettoPoint::vartime_multiscalar_mul(
+            iter::once(x) // A_I
+                .chain(iter::once(x * x)) // A_O
+                .chain(iter::once(x)) // W_L_point
+                .chain(iter::once(x)) // W_R_point
+                .chain(iter::once(Scalar::one())) // W_O_point
+                .chain(iter::once(x * x * x)) // S
+                .chain(iter::once(w * (self.t_x - a * b) + r * (x * x * (delta + z_c) - self.t_x))) // B
+                .chain(iter::once(-self.e_blinding - r * self.t_x_blinding)) // B_blinding
+                .chain(g) // G
+                .chain(h) // H
+                .chain(x_sq.iter().cloned()) // ipp_proof.L_vec
+                .chain(x_inv_sq.iter().cloned()) // ipp_proof.R_vec
+                .chain(V_multiplier) // V
+                .chain(iter::once(r * x)) // T_1
+                .chain(iter::once(r * x * x * x)) // T_3
+                .chain(iter::once(r * x * x * x * x)) // T_4
+                .chain(iter::once(r * x * x * x * x * x)) // T_5
+                .chain(iter::once(r * x * x * x * x * x * x)), // T_6
+            iter::once(&self.A_I)
+                .chain(iter::once(&self.A_O))
+                .chain(iter::once(&W_L_point))
+                .chain(iter::once(&W_R_point))
+                .chain(iter::once(&W_O_point))
+                .chain(iter::once(&self.S))
+                .chain(iter::once(&gen.pedersen_generators.B))
+                .chain(iter::once(&gen.pedersen_generators.B_blinding))
+                .chain(gen.G.iter())
+                .chain(gen.H.iter())
+                .chain(self.ipp_proof.L_vec.iter())
+                .chain(self.ipp_proof.R_vec.iter())
+                .chain(V.iter())
+                .chain(iter::once(&self.T_1))
+                .chain(iter::once(&self.T_3))
+                .chain(iter::once(&self.T_4))
+                .chain(iter::once(&self.T_5))
+                .chain(iter::once(&self.T_6)),
+        );
```

Comment:
I think this equation could be simplified and clarified by breaking its components up into pieces and then joining the pieces (like `t_scalars, T_points`, etc).

Reply:
I joined the `T_scalars` and `T_points`, that's a good idea.
I think any further simplification / flattening would make it harder to follow along with the math though, since the current version is a pretty direct reflection of the math in the paper/notes.